client/topology.go

```diff
@@ -0,0 +1,245 @@
+		if v.equals(tuple) {
+			return true
+		}
+	}
+	return false
+}
+
+// TopologyService is a helper service to combine data from multiple API calls into a single data presentation for the frontend
+type TopologyService struct {
+	HTTPClient HTTPClient
+}
+
+// NetworkTopology returns data about connected peers and "friends of friends".
+func (ts TopologyService) NetworkTopology(ctx context.Context) (NetworkTopology, error) {
+	// first get diagnostics for our node's information
+	networkTopology := NetworkTopology{
+		Edges: []Tuple{},
+	}
+	diagnostics, err := ts.HTTPClient.Diagnostics(ctx)
+	if err != nil {
+		return networkTopology, err
+	}
+	networkTopology.PeerID = diagnostics.Network.Connections.PeerId
+	networkTopology.TxCount = diagnostics.Network.State.TransactionCount
+
+	// Add own as vertice, needed for empty node
+	networkTopology.Peers = append(networkTopology.Peers, Peer{PeerID: diagnostics.Network.Connections.PeerId, TransactionCount: diagnostics.Network.State.TransactionCount})
+
+	// Peer diagnostics, this will include connections from our node to others as well.
+	peerDiagnostics, err := ts.HTTPClient.PeerDiagnostics(ctx)
+	if err != nil {
+		return networkTopology, err
+	}
+
+	for k, v := range peerDiagnostics {
+		peerID := realPeerID(k)
+		peer, ok := existingPeer(networkTopology.Peers, peerID)
+		if v.TransactionNum != nil {
+			peer.TransactionCount = int(*v.TransactionNum)
+		}
+		if v.SoftwareVersion != nil {
+			peer.SoftwareVersion = *v.SoftwareVersion
+		}
+		if v.SoftwareID != nil {
+			peer.SoftwareID = *v.SoftwareID
+		}
+		if !ok {
+			networkTopology.Peers = append(networkTopology.Peers, *peer)
+		}
+		if v.Peers != nil {
+			for _, connectedPeer := range *v.Peers {
+				otherPeerID := realPeerID(connectedPeer)
+				t := Tuple([2]string{otherPeerID, peerID})
+				if !containsEdge(networkTopology.Edges, t) {
+					networkTopology.Edges = append(networkTopology.Edges, t)
+				}
+				if p, ok := existingPeer(networkTopology.Peers, otherPeerID); !ok {
+					networkTopology.Peers = append(networkTopology.Peers, *p)
+				}
+			}
+		}
+	}
+
+	// add data to peers we are connected to
+	for _, cp := range diagnostics.Network.Connections.ConnectedPeers {
+		peer, _ := existingPeer(networkTopology.Peers, cp.Id)
+		peer.NodeDID = cp.Nodedid
+		peer.Address = cp.Address
+		peer.Authenticated = cp.Authenticated
+	}
+
+	// this is a blocking call that opens TLS connections to all peers
+	ts.addInfoToPeers(ctx, networkTopology.Peers)
+
+	return networkTopology, nil
+}
+
+// addInfoToPeers adds information from DID Documents and the Certificate exposed at the NutsComm address.
+func (ts TopologyService) addInfoToPeers(ctx context.Context, peers []Peer) {
+	wgDocument := sync.WaitGroup{}
+	wgTLS := sync.WaitGroup{}
+	for i, p := range peers {
+		if p.Authenticated {
+			// connect to the node and certificate info
+			wgDocument.Add(1)
+			go func(peer *Peer) {
+				document, err := ts.HTTPClient.DIDDocument(ctx, *peer.NodeDID)
```

Comment:
You might want to cache this, since this is an N+1 type query?
(same for the TLS dial)